File: ts/indicators/john-bollinger/bollinger-bands/bollinger-bands.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, DefaultBarComponent, barComponentValue } from '../../../entities/bar-component';
import { Quote } from '../../../entities/quote';
import { QuoteComponent, DefaultQuoteComponent, quoteComponentValue } from '../../../entities/quote-component';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { TradeComponent, DefaultTradeComponent, tradeComponentValue } from '../../../entities/trade-component';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { Band } from '../../core/outputs/band';
import { ExponentialMovingAverage } from '../../common/exponential-moving-average/exponential-moving-average';
import { SimpleMovingAverage } from '../../common/simple-moving-average/simple-moving-average';
import { Variance } from '../../common/variance/variance';
import { BollingerBandsParams, MovingAverageType } from './params';

/** Interface for an indicator that accepts a scalar and returns a value. */
interface LineUpdater {
  update(sample: number): number;
  isPrimed(): boolean;
}

/** Function to calculate mnemonic of a __BollingerBands__ indicator. */
export const bollingerBandsMnemonic = (length: number, upperMultiplier: number, lowerMultiplier: number,
  barComponent?: BarComponent, quoteComponent?: QuoteComponent, tradeComponent?: TradeComponent): string => {
  const cm = componentTripleMnemonic(barComponent, quoteComponent, tradeComponent);
  return `bb(${length},${Math.floor(upperMultiplier)},${Math.floor(lowerMultiplier)}${cm})`;
};

/**
 * John Bollinger's Bollinger Bands indicator.
 *
 * Bollinger Bands consist of a middle band (moving average) and upper/lower bands
 * placed a specified number of standard deviations above and below the middle band.
 *
 * The indicator produces six outputs:
 *   - LowerValue: middleValue - lowerMultiplier * stddev
 *   - MiddleValue: moving average of the input
 *   - UpperValue: middleValue + upperMultiplier * stddev
 *   - BandWidth: (upperValue - lowerValue) / middleValue
 *   - PercentBand: (sample - lowerValue) / (upperValue - lowerValue)
 *   - Band: lower/upper band pair
 *
 * Reference:
 *
 * Bollinger, John (2002). Bollinger on Bollinger Bands. McGraw-Hill.
 */
export class BollingerBands implements Indicator {

  private readonly ma: LineUpdater;
  private readonly variance_: Variance;
  private readonly upperMultiplier_: number;
  private readonly lowerMultiplier_: number;

  private readonly barComponentFunc: (bar: Bar) => number;
  private readonly quoteComponentFunc: (quote: Quote) => number;
  private readonly tradeComponentFunc: (trade: Trade) => number;

  private middleValue_ = NaN;
  private upperValue_ = NaN;
  private lowerValue_ = NaN;
  private bandWidth_ = NaN;
  private percentBand_ = NaN;
  private primed_ = false;

  private readonly mnemonic_: string;
  private readonly description_: string;

  constructor(params: BollingerBandsParams) {
    const length = Math.floor(params.length);

    if (length < 2) {
      throw new Error('length should be greater than 1');
    }

    const upperMultiplier = params.upperMultiplier ?? 2.0;
    const lowerMultiplier = params.lowerMultiplier ?? 2.0;
    const isUnbiased = params.isUnbiased ?? true;

    this.upperMultiplier_ = upperMultiplier;
    this.lowerMultiplier_ = lowerMultiplier;

    const bc = params.barComponent ?? DefaultBarComponent;
    const qc = params.quoteComponent ?? DefaultQuoteComponent;
    const tc = params.tradeComponent ?? DefaultTradeComponent;

    this.barComponentFunc = barComponentValue(bc);
    this.quoteComponentFunc = quoteComponentValue(qc);
    this.tradeComponentFunc = tradeComponentValue(tc);

    // Create variance sub-indicator.
    this.variance_ = new Variance({
      length,
      unbiased: isUnbiased,
      barComponent: params.barComponent,
      quoteComponent: params.quoteComponent,
      tradeComponent: params.tradeComponent,
    });

    // Create moving average sub-indicator.
    if (params.movingAverageType === MovingAverageType.EMA) {
      this.ma = new ExponentialMovingAverage({
        length,
        firstIsAverage: params.firstIsAverage ?? false,
      });
    } else {
      this.ma = new SimpleMovingAverage({ length });
    }

    this.mnemonic_ = bollingerBandsMnemonic(length, upperMultiplier, lowerMultiplier,
      params.barComponent, params.quoteComponent, params.tradeComponent);
    this.description_ = 'Bollinger Bands ' + this.mnemonic_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.primed_;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.BollingerBands,
      this.mnemonic_,
      this.description_,
      [
        { mnemonic: this.mnemonic_ + ' lower', description: this.description_ + ' Lower' },
        { mnemonic: this.mnemonic_ + ' middle', description: this.description_ + ' Middle' },
        { mnemonic: this.mnemonic_ + ' upper', description: this.description_ + ' Upper' },
        { mnemonic: this.mnemonic_ + ' bandWidth', description: this.description_ + ' Band Width' },
        { mnemonic: this.mnemonic_ + ' percentBand', description: this.description_ + ' Percent Band' },
        { mnemonic: this.mnemonic_ + ' band', description: this.description_ + ' Band' },
      ],
    );
  }

  /**
   * Updates the indicator given the next sample value.
   * Returns [lower, middle, upper, bandWidth, percentBand].
   */
  public update(sample: number): [number, number, number, number, number] {
    if (isNaN(sample)) {
      return [NaN, NaN, NaN, NaN, NaN];
    }

    const middle = this.ma.update(sample);
    const v = this.variance_.update(sample);

    this.primed_ = this.ma.isPrimed() && this.variance_.isPrimed();

    if (isNaN(middle) || isNaN(v)) {
      this.middleValue_ = NaN;
      this.upperValue_ = NaN;
      this.lowerValue_ = NaN;
      this.bandWidth_ = NaN;
      this.percentBand_ = NaN;
      return [NaN, NaN, NaN, NaN, NaN];
    }

    const stddev = Math.sqrt(v);
    const upper = middle + this.upperMultiplier_ * stddev;
    const lower = middle - this.lowerMultiplier_ * stddev;

    const epsilon = 1e-10;

    let bw: number;
    if (Math.abs(middle) < epsilon) {
      bw = 0;
    } else {
      bw = (upper - lower) / middle;
    }

    let pctB: number;
    const spread = upper - lower;
    if (Math.abs(spread) < epsilon) {
      pctB = 0;
    } else {
      pctB = (sample - lower) / spread;
    }

    this.middleValue_ = middle;
    this.upperValue_ = upper;
    this.lowerValue_ = lower;
    this.bandWidth_ = bw;
    this.percentBand_ = pctB;

    return [lower, middle, upper, bw, pctB];
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const [lower, middle, upper, bw, pctB] = this.update(sample.value);

    const s0 = new Scalar(); s0.time = sample.time; s0.value = lower;
    const s1 = new Scalar(); s1.time = sample.time; s1.value = middle;
    const s2 = new Scalar(); s2.time = sample.time; s2.value = upper;
    const s3 = new Scalar(); s3.time = sample.time; s3.value = bw;
    const s4 = new Scalar(); s4.time = sample.time; s4.value = pctB;

    const band = new Band();
    band.time = sample.time;
    if (isNaN(lower) || isNaN(upper)) {
      band.lower = NaN;
      band.upper = NaN;
    } else {
      band.lower = lower;
      band.upper = upper;
    }

    return [s0, s1, s2, s3, s4, band];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const v = this.barComponentFunc(sample);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = this.quoteComponentFunc(sample);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const v = this.tradeComponentFunc(sample);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }
}

File: go/indicators/johnbollinger/bollingerbands/bollingerbands.go
package bollingerbands

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/common/exponentialmovingaverage"
	"zpano/indicators/common/simplemovingaverage"
	"zpano/indicators/common/variance"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// lineUpdater is an interface for indicators that accept a single scalar and return a value.
type lineUpdater interface {
	Update(float64) float64
	IsPrimed() bool
}

// BollingerBands is John Bollinger's Bollinger Bands indicator.
//
// Bollinger Bands consist of a middle band (moving average) and upper/lower bands
// placed a specified number of standard deviations above and below the middle band.
//
// The indicator produces six outputs:
//   - LowerValue: middleValue - lowerMultiplier * stddev
//   - MiddleValue: moving average of the input
//   - UpperValue: middleValue + upperMultiplier * stddev
//   - BandWidth: (upperValue - lowerValue) / middleValue
//   - PercentBand: (sample - lowerValue) / (upperValue - lowerValue)
//   - Band: lower/upper band pair
//
// Reference:
//
// Bollinger, John (2002). Bollinger on Bollinger Bands. McGraw-Hill.
type BollingerBands struct {
	mu sync.RWMutex

	ma       lineUpdater
	variance *variance.Variance

	upperMultiplier float64
	lowerMultiplier float64

	middleValue float64
	upperValue  float64
	lowerValue  float64
	bandWidth   float64
	percentBand float64
	primed      bool

	barFunc   entities.BarFunc
	quoteFunc entities.QuoteFunc
	tradeFunc entities.TradeFunc

	mnemonic string
}

// NewBollingerBands returns an instance of the indicator created using supplied parameters.
//
//nolint:funlen,cyclop
func NewBollingerBands(p *BollingerBandsParams) (*BollingerBands, error) {
	const (
		invalid          = "invalid bollinger bands parameters"
		fmts             = "%s: %s"
		fmtw             = "%s: %w"
		minLength        = 2
		defaultLength    = 5
		defaultMultipler = 2.0
	)

	length := p.Length
	if length == 0 {
		length = defaultLength
	}

	if length < minLength {
		return nil, fmt.Errorf(fmts, invalid, "length should be greater than 1")
	}

	upperMultiplier := p.UpperMultiplier
	if upperMultiplier == 0 {
		upperMultiplier = defaultMultipler
	}

	lowerMultiplier := p.LowerMultiplier
	if lowerMultiplier == 0 {
		lowerMultiplier = defaultMultipler
	}

	isUnbiased := true
	if p.IsUnbiased != nil {
		isUnbiased = *p.IsUnbiased
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Create variance sub-indicator.
	v, err := variance.NewVariance(&variance.VarianceParams{
		Length:         length,
		IsUnbiased:     isUnbiased,
		BarComponent:   bc,
		QuoteComponent: qc,
		TradeComponent: tc,
	})
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Create moving average sub-indicator.
	var ma lineUpdater

	var maLabel string

	switch p.MovingAverageType {
	case EMA:
		maLabel = "EMA"

		ema, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         length,
				FirstIsAverage: p.FirstIsAverage,
			})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		ma = ema
	default:
		maLabel = "SMA"

		sma, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: length})
		if e != nil {
			return nil, fmt.Errorf(fmtw, invalid, e)
		}

		ma = sma
	}

	_ = maLabel // Reserved for future mnemonic use.

	mnemonic := fmt.Sprintf("bb(%d,%.0f,%.0f%s)", length, upperMultiplier, lowerMultiplier,
		core.ComponentTripleMnemonic(bc, qc, tc))

	return &BollingerBands{
		ma:              ma,
		variance:        v,
		upperMultiplier: upperMultiplier,
		lowerMultiplier: lowerMultiplier,
		middleValue:     math.NaN(),
		upperValue:      math.NaN(),
		lowerValue:      math.NaN(),
		bandWidth:       math.NaN(),
		percentBand:     math.NaN(),
		barFunc:         barFunc,
		quoteFunc:       quoteFunc,
		tradeFunc:       tradeFunc,
		mnemonic:        mnemonic,
	}, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *BollingerBands) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *BollingerBands) Metadata() core.Metadata {
	desc := "Bollinger Bands " + s.mnemonic

	return core.BuildMetadata(
		core.BollingerBands,
		s.mnemonic,
		desc,
		[]core.OutputText{
			{Mnemonic: s.mnemonic + " lower", Description: desc + " Lower"},
			{Mnemonic: s.mnemonic + " middle", Description: desc + " Middle"},
			{Mnemonic: s.mnemonic + " upper", Description: desc + " Upper"},
			{Mnemonic: s.mnemonic + " bandWidth", Description: desc + " Band Width"},
			{Mnemonic: s.mnemonic + " percentBand", Description: desc + " Percent Band"},
			{Mnemonic: s.mnemonic + " band", Description: desc + " Band"},
		},
	)
}

// Update updates the indicator given the next sample value.
// Returns lower, middle, upper, bandWidth, percentBand values.
//
//nolint:nonamedreturns
func (s *BollingerBands) Update(sample float64) (lower, middle, upper, bw, pctB float64) {
	nan := math.NaN()

	if math.IsNaN(sample) {
		return nan, nan, nan, nan, nan
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	middle = s.ma.Update(sample)
	v := s.variance.Update(sample)

	s.primed = s.ma.IsPrimed() && s.variance.IsPrimed()

	if math.IsNaN(middle) || math.IsNaN(v) {
		s.middleValue = nan
		s.upperValue = nan
		s.lowerValue = nan
		s.bandWidth = nan
		s.percentBand = nan

		return nan, nan, nan, nan, nan
	}

	stddev := math.Sqrt(v)
	upper = middle + s.upperMultiplier*stddev
	lower = middle - s.lowerMultiplier*stddev

	const epsilon = 1e-10

	if math.Abs(middle) < epsilon {
		bw = 0
	} else {
		bw = (upper - lower) / middle
	}

	spread := upper - lower
	if math.Abs(spread) < epsilon {
		pctB = 0
	} else {
		pctB = (sample - lower) / spread
	}

	s.middleValue = middle
	s.upperValue = upper
	s.lowerValue = lower
	s.bandWidth = bw
	s.percentBand = pctB

	return lower, middle, upper, bw, pctB
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *BollingerBands) UpdateScalar(sample *entities.Scalar) core.Output {
	lower, middle, upper, bw, pctB := s.Update(sample.Value)

	const outputCount = 6

	output := make([]any, outputCount)
	output[0] = entities.Scalar{Time: sample.Time, Value: lower}
	output[1] = entities.Scalar{Time: sample.Time, Value: middle}
	output[2] = entities.Scalar{Time: sample.Time, Value: upper}
	output[3] = entities.Scalar{Time: sample.Time, Value: bw}
	output[4] = entities.Scalar{Time: sample.Time, Value: pctB}

	if math.IsNaN(lower) || math.IsNaN(upper) {
		output[5] = outputs.NewEmptyBand(sample.Time)
	} else {
		output[5] = outputs.NewBand(sample.Time, lower, upper)
	}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (s *BollingerBands) UpdateBar(sample *entities.Bar) core.Output {
	v := s.barFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *BollingerBands) UpdateQuote(sample *entities.Quote) core.Output {
	v := s.quoteFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *BollingerBands) UpdateTrade(sample *entities.Trade) core.Output {
	v := s.tradeFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}
